courses/models_test.py

```diff
@@ -15,13 +16,56 @@
     ProgramEnrollmentFactory,
     ProgramFactory,
 )
-from courses.models import CourseRunEnrollment
+from courses.models import Course, CourseRunEnrollment, ProgramRequirement
 from main.test_utils import format_as_iso8601
 from users.factories import UserFactory
 
 pytestmark = [pytest.mark.django_db]
 
 
+@pytest.fixture
+def program_with_requirements():
+    program = ProgramFactory.create()
+    required_courses = CourseFactory.create_batch(3, program=program)
+    elective_courses = CourseFactory.create_batch(3, program=program)
+    mut_exclusive_courses = CourseFactory.create_batch(3, program=program)
```

Comment:
I'm not sure I understand how mutually exclusive works. 

In DEDP, if you take both 14.73x and 14.009x you will only get credit towards the program certificate for one of them. 

So would this be 

```suggestion
    mut_exclusive_courses = CourseFactory.create_batch(2, program=program)
```

Reply:
So to summarize the logic here in psuedocode:
```python
has_cert = (
   # required
   # ALL_OF
   all(
      # must take all of these
      [r.passed for r in required_courses]
   )
   # at least min_elective_count electives passed
   # MIN_NUM_OF, value=5
   and sum(
      [1 for r in elective_courses if r.passed] +
      # mutually exclusive, only 1 counts
       # MIN_NUM_OF, value=1
      [1 if any([r.passed for r in mut_exclusive_courses]) else 0]
   ) > 5
)
```

The structure here is very flexible, it can support fairly complex things like this (I've seen things like this in residential programs):

```python
has_cert = (
   # ALL_OF
   all(
       # must take all of these
       [r.passed for r in required_courses] + 
       # any of these, but you MUST take one
       # MIN_NUM_OF, value=1
       [any([r.passed for r in mut_exclusive_required_courses1])]
       # any of these, but you MUST take one
       # MIN_NUM_OF, value=1
       [any([r.passed for r in mut_exclusive_required_courses2])]
   )
   # at least min_elective_count electives passed
   # MIN_NUM_OF, value=5
   and sum(
      [1 for r in elective_courses if r.passed] +
      # mutually exclusive, only 1 counts
      # MIN_NUM_OF, value=1
      [1 if any([r.passed for r in mut_exclusive_courses1]) else 0]
      # mutually exclusive, only 1 counts
      # MIN_NUM_OF, value=1
      [1 if any([r.passed for r in mut_exclusive_courses2]) else 0]
   ) > 5
)
```

Reply:
Very broadly, the data model here representation a logical tree of boolean checks and for MIN_NUM_OF a check of a minimum threshold being met.

Reply:
A more concrete example is probably more useful to illustrate:

```python
program = ProgramFactory.create(name="DEDP")

root_node = ProgramRequirement.add_root(program=program)

required_courses_node = root_node.add_child(
    operator=ProgramRequirement.Operator.ALL_OF, title="Required Courses"
)
for required in [
    CourseFactory.create(readable_id="course-v1:MITxT+14.100x"),
    CourseFactory.create(readable_id="course-v1:MITxT+JPAL102x"),
    CourseFactory.create(readable_id="course-v1:MITxT+14.310x"),
]:
    required_courses_node.add_child(course=required)

elective_courses_node = root_node.add_child(
    operator=ProgramRequirement.Operator.MIN_NUMBER_OF,
    operator_value=2,
    title="Elective Courses",
)
for elective in [
    CourseFactory.create(readable_id="course-v1:MITxT+14.750x"),
    CourseFactory.create(readable_id="course-v1:MITxT+14.740x"),
    [
        CourseFactory.create(readable_id="course-v1:MITxT+14.73x"),
        CourseFactory.create(readable_id="course-v1:MITxT+14.009x"),
    ]
]:
    if isinstance(elective, list):
        elective_courses_node.add_child(
            operator=ProgramRequirement.Operator.MIN_NUMBER_OF, operator_value=1
        )
    else:
        elective_courses_node.add_child(course=elective)
```

Reply:
Ok. That wasn't what I was expecting, but it makes sense to me, with the tree structure. 

What does this look like through the django admin? How do we add or change requirements?

Reply:
Good question on how to add or change requirements, so a question for you:

Will we have the need (similar to certificates) to display what the historical requirements were when a learner earned a certificate?